### ISSUE TITLE
Missing components / tests on Windows

### ISSUE THREAD

**Issue by yf225** [pytorch/pytorch#4092]
This issue tracks the components / tests that are not working on Windows:

- [ ] `test/test_distributed.py`: currently disabled on Windows because Windows doesn't have `fcntl` and we need to look for substitutes
- [ ] `torch/csrc/jit/fusion_compiler.cpp`: Fuser is disabled on Windows because the current implementation uses symbols from Linux-specific headers `dlfcn.h` and `unistd.h`. We will need to find alternatives for Windows.
- [ ] `test/test_torch.py`: some parts of `test_serialization` and `test_from_file` are disabled because Windows doesn't support opening an already opened file (see discussion at https://github.com/pytorch/pytorch/pull/4065/files/4a7dfbd12c77850faa4171235b18b9b1540f3f0d#diff-9996665f82f52030836eb8657057cfadR4227 and https://github.com/pytorch/pytorch/pull/4065/files/4a7dfbd12c77850faa4171235b18b9b1540f3f0d#diff-9996665f82f52030836eb8657057cfadR4398)
- [ ] `test/test_nccl`: `nccl` currently doesn't work with Windows (https://github.com/NVIDIA/nccl/pull/31 is the porting diff)
- [ ] `test/test_utils.py`: `test_cuda` in `TestBottleneck` causes intermittent CUDA out-of-memory error on Windows 
- [ ] `test/test_utils.py`, `test/test_dataloader.py`: DataLoader with multiple workers causes intermittent CUDA out-of-memory error on Windows.
- [x] ~~`test/test_cpp_extensions.py`: this test currently doesn't work on Windows~~
- [x] ~~`test/test_dataloader.py`, `test/test_multiprocessing.py`: various tests are disabled because there are issues when using `multiprocessing` on Windows~~ (done in https://github.com/pytorch/pytorch/pull/4453)
- [x] ~~`test/test_autograd.py`: `test_profiler` is disabled because Windows doesn't have `c++filt` (see discussion at https://github.com/pytorch/pytorch/pull/4065/files/4a7dfbd12c77850faa4171235b18b9b1540f3f0d#diff-e7525582ab432c30821289d64cc52bc3R1653)~~
- [x] ~~#4615 causes mysterious failures ("RuntimeError: Caught an unknown exception!") in 3 JIT tests~~
  - ~~`test_arg_configurations`~~
  - ~~`test_multiuse_fn`~~
  - ~~`test_output_unflatten`~~
- [x] ~~`test/test_torch.py`, `test/test_nn.py`: CUDA HalfTensor currently doesn't work on Windows~~
- [x] ~~`test/test_nn.py`: `test_Conv2d_deterministic_cudnn` seems to be stuck on Windows~~
- [x] ~~`test/test_cuda.py`:~~
   - ~~Some tests are turned off because CUDA HalfTensor currently doesn't work on Windows. Test logs to look at:~~
~~https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-test/57/console~~
   - ~~`test_FloatTensor_qr_big` and `test_DoubleTensor_qr_big` are known to stuck or have numerical errors intermittently~~

For more discussions, also see: https://github.com/pytorch/pytorch/pull/2941
  
  

cc @mruberry @peterjc123 @nbcsm @guyang3532

**Comment by peterjc123** [comment]
@yf225 The first one is solved by #4453 but I think DataLoader can be further improved when #4000 is fininshed.

**Comment by yf225** [comment]
@peterjc123 Cool I will mark it as resolved :)

**Comment by peterjc123** [comment]
@yf225 The `test_profiler` has been added in #4522.

**Comment by peterjc123** [comment]
@yf225 I think we can try to revert #5591 and #5322 once #5590 is merged. Since these memory leaks are limited to Windows, and the CUDA Tensors are not shared across processes, I think the leak may be related to this PR.

**Comment by peterjc123** [comment]
@yf225 We can try to revert #5591 and #5322, because the memory leak in the CPU side could also cause CUDA errors.

**Comment by yf225** [comment]
@peterjc123 Are they fixed by https://github.com/pytorch/pytorch/pull/5585? I think we can revert them after we merge the PR.

**Comment by peterjc123** [comment]
@yf225 I think this may be related. Since once the memory of the CPU side is low, the `cudaMalloc` will fail with `cuda error(2): out of memory` too.

**Comment by yf225** [comment]
@peterjc123 I think it could be. For what's worth, when I tried to inspect the CUDA OOM error, `nvidia-smi` showed no process that was taking memory, but running CUDA tests on the machine would still fail.

**Comment by peterjc123** [comment]
@yf225 Now that #5585 is merged into master, could you please try to revert the changes on #5591 and #5322?

**Comment by yf225** [comment]
@peterjc123 Awesome! Just to understand it better: does #5585 fix both num_worker=1 and num_worker>1 cases?

**Comment by peterjc123** [comment]
@yf225 Yes, they are both solved.

**Comment by peterjc123** [comment]
@yf225 I guess we should mark the cpp_extension test as completed, for it's now enabled in CI.

**Comment by mruberry** [comment]
Closing this issue due to age and because its references are long out of date. For example, distributed tests now have their own jobs.
